File: ScholarshipEligibilityEvaluator.py
from dataclasses import dataclass
from enum import Enum
from typing import List


class Status(Enum):
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    MANUAL_REVIEW = "MANUAL_REVIEW"


@dataclass(frozen=True)
class EvaluationResult:
    status: Status
    reasons: List[str]


def evaluate_scholarship(
    age: int,
    gpa: float,
    attendance_rate: float,
    has_required_courses: bool,
    disciplinary_record: bool
) -> EvaluationResult:
    _validate_inputs(gpa, attendance_rate)

    rejection_reasons = []
    review_reasons = []

    # Rule 1 - Age
    if age < 16:
        rejection_reasons.append("Applicant is younger than the minimum age.")
    elif age <= 17:
        review_reasons.append("Applicant is under 18 and requires manual review.")

    # Rule 2 - GPA
    if gpa < 6.0:
        rejection_reasons.append("GPA is below the minimum required.")
    elif gpa < 7.0:
        review_reasons.append("GPA is in the manual review range.")

    # Rule 3 - Attendance
    if attendance_rate < 75.0:
        rejection_reasons.append("Attendance rate is below the minimum required.")
    elif attendance_rate < 80.0:
        review_reasons.append("Attendance rate is in the manual review range.")

    # Rule 4 - Required courses
    if not has_required_courses:
        rejection_reasons.append("Required courses have not been completed.")

    # Rule 5 - Disciplinary record
    if disciplinary_record:
        rejection_reasons.append("Applicant has a disciplinary record.")

    # Final decision
    if rejection_reasons:
        return EvaluationResult(
            status=Status.REJECTED,
            reasons=rejection_reasons
        )

    if review_reasons:
        return EvaluationResult(
            status=Status.MANUAL_REVIEW,
            reasons=review_reasons
        )

    return EvaluationResult(
        status=Status.APPROVED,
        reasons=["Applicant meets all scholarship requirements."]
    )


def _validate_inputs(gpa: float, attendance_rate: float) -> None:
    if gpa < 0.0 or gpa > 10.0:
        raise ValueError("GPA must be between 0 and 10.")

    if attendance_rate < 0.0 or attendance_rate > 100.0:
        raise ValueError("Attendance rate must be between 0 and 100.")


if __name__ == "__main__":
    result = evaluate_scholarship(
        age=18,
        gpa=8.5,
        attendance_rate=92.0,
        has_required_courses=True,
        disciplinary_record=False
    )
    print(result)

File: ScholarshipEligibilityEvaluator.java
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ScholarshipEligibilityEvaluator {

    public enum Status {
        APPROVED,
        REJECTED,
        MANUAL_REVIEW
    }

    public static class EvaluationResult {
        private final Status status;
        private final List<String> reasons;

        public EvaluationResult(Status status, List<String> reasons) {
            this.status = Objects.requireNonNull(status);
            this.reasons = List.copyOf(reasons);
        }

        public Status getStatus() {
            return status;
        }

        public List<String> getReasons() {
            return reasons;
        }

        @Override
        public String toString() {
            return "EvaluationResult{" +
                    "status=" + status +
                    ", reasons=" + reasons +
                    '}';
        }
    }

    public static EvaluationResult evaluateScholarship(
            int age,
            double gpa,
            double attendanceRate,
            boolean hasRequiredCourses,
            boolean disciplinaryRecord
    ) {
        validateInputs(gpa, attendanceRate);

        List<String> rejectionReasons = new ArrayList<>();
        List<String> reviewReasons = new ArrayList<>();

        // Rule 1 - Age
        if (age < 16) {
            rejectionReasons.add("Applicant is younger than the minimum age.");
        } else if (age <= 17) {
            reviewReasons.add("Applicant is under 18 and requires manual review.");
        }

        // Rule 2 - GPA
        if (gpa < 6.0) {
            rejectionReasons.add("GPA is below the minimum required.");
        } else if (gpa < 7.0) {
            reviewReasons.add("GPA is in the manual review range.");
        }

        // Rule 3 - Attendance
        if (attendanceRate < 75.0) {
            rejectionReasons.add("Attendance rate is below the minimum required.");
        } else if (attendanceRate < 80.0) {
            reviewReasons.add("Attendance rate is in the manual review range.");
        }

        // Rule 4 - Required courses
        if (!hasRequiredCourses) {
            rejectionReasons.add("Required courses have not been completed.");
        }

        // Rule 5 - Disciplinary record
        if (disciplinaryRecord) {
            rejectionReasons.add("Applicant has a disciplinary record.");
        }

        // Final decision
        if (!rejectionReasons.isEmpty()) {
            return new EvaluationResult(Status.REJECTED, rejectionReasons);
        }

        if (!reviewReasons.isEmpty()) {
            return new EvaluationResult(Status.MANUAL_REVIEW, reviewReasons);
        }

        return new EvaluationResult(Status.APPROVED, List.of("Applicant meets all scholarship requirements."));
    }

    private static void validateInputs(double gpa, double attendanceRate) {
        if (gpa < 0.0 || gpa > 10.0) {
            throw new IllegalArgumentException("GPA must be between 0 and 10.");
        }

        if (attendanceRate < 0.0 || attendanceRate > 100.0) {
            throw new IllegalArgumentException("Attendance rate must be between 0 and 100.");
        }
    }

    public static void main(String[] args) {
        EvaluationResult result = evaluateScholarship(
                18,
                8.5,
                92.0,
                true,
                false
        );

        System.out.println(result);
    }
}